ecs-init/cache/cache.go

```diff
@@ -0,0 +1,173 @@
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Package cache provides functionality for working with an on-disk cache of
+// the ECS Agent image.
+package cache
+
+import (
+	"crypto/md5"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+
+	"github.com/aws/amazon-ecs-init/ecs-init/config"
+	log "github.com/cihub/seelog"
+)
+
+// Downloader is resposible for cache operations relating to downloading the agent
+type Downloader struct {
+	getter httpGetter
+	fs     fileSystem
+}
+
+// NewDownloader returns a Downloader with default dependencies
+func NewDownloader() *Downloader {
+	return &Downloader{
+		getter: customGetter,
+		fs:     &standardFS{},
+	}
+}
+
+// IsAgentCached returns true if there is a cached copy of the Agent present
+// (no validation is performed)
+func (d *Downloader) IsAgentCached() bool {
+	file, err := d.fs.Open(config.AgentTarball())
+	if err != nil {
+		return false
+	}
+	file.Close()
+	return true
+}
+
+// IsAgentLatest checks whether the cached copy of the Agent has the same MD5
+// sum as the published MD5 sum
+func (d *Downloader) IsAgentLatest() bool {
+	file, err := d.fs.Open(config.AgentTarball())
+	if err != nil {
+		return false
+	}
+	defer file.Close()
+
+	md5hash := md5.New()
+	_, err = d.fs.Copy(md5hash, file)
+	if err != nil {
+		log.Error("Could not calculate md5sum", err)
+		return false
+	}
+
+	publishedMd5Sum, err := d.getPublishedMd5Sum()
+	if err != nil {
+		return false
+	}
+
+	calculatedMd5Sum := md5hash.Sum(nil)
+	calculatedMd5SumString := fmt.Sprintf("%x", calculatedMd5Sum)
+	log.Debugf("Expected %s", publishedMd5Sum)
+	log.Debugf("Calculated %s", calculatedMd5SumString)
+	if publishedMd5Sum != calculatedMd5SumString {
+		log.Info("Cached Amazon EC2 Container Service Agent does not match latest at %s", config.AgentRemoteTarball)
+		return false
+	}
+	return true
+}
+
+// LoadCachedAgent returns an io.ReadCloser of the Agent from the cache
+func (d *Downloader) LoadCachedAgent() (io.ReadCloser, error) {
+	return d.fs.Open(config.AgentTarball())
+}
+
+// DownloadAgent downloads a fresh copy of the Agent and performs an
+// integrity check on the downloaded image
+func (d *Downloader) DownloadAgent() error {
+	err := d.fs.MkdirAll(config.CacheDirectory(), os.ModeDir|0700)
+	if err != nil {
+		return err
+	}
+
+	publishedMd5Sum, err := d.getPublishedMd5Sum()
+	if err != nil {
+		return err
+	}
+
+	publishedTarballReader, err := d.getPublishedTarball()
+	if err != nil {
+		return err
+	}
+	defer publishedTarballReader.Close()
+
+	md5hash := md5.New()
+	tempFile, err := d.fs.TempFile("", "ecs-agent.tar")
+	if err != nil {
+		return err
+	}
+	defer tempFile.Close()
```

Comment:
You probably want to swap the order of your two defers so that you close then delete the file, not the reverse.